src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,237 @@
+#include "Fingerprint.hpp"
+#include "../helpers/Log.hpp"
+#include "src/config/ConfigManager.hpp"
+
+#include <filesystem>
+#include <unistd.h>
+#include <pwd.h>
+#include <security/pam_appl.h>
+#if __has_include(<security/pam_misc.h>)
+#include <security/pam_misc.h>
+#endif
+
+#include <cstring>
+#include <thread>
+
+static const auto FPRINT        = "net.reactivated.Fprint";
+static const auto DEVICE        = "net.reactivated.Fprint.Device";
+static const auto MANAGER       = "net.reactivated.Fprint.Manager";
+static const auto LOGIN_MANAGER = "org.freedesktop.login1.Manager";
+static const auto RETRY_MESSAGE = "Could not match fingerprint. Try again.";
+
+CFingerprint::CFingerprint() {}
+
+static void fingerprintCheckTimerCallback(std::shared_ptr<CTimer> self, void* data) {
+    g_pHyprlock->unlock();
+}
+
+void CFingerprint::start() {
+    static auto* const PENABLEFINGERPRINT = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:enable_fingerprint");
+    const bool         ENABLEFINGERPRINT  = **PENABLEFINGERPRINT;
+    if (!ENABLEFINGERPRINT) {
+        return;
+    }
+    std::thread([this]() {
+        // For grace or SIGUSR1 unlocks
+        if (g_pHyprlock->isUnlocked())
+            return;
+
+        setupDBUS();
+
+        g_pHyprlock->addTimer(std::chrono::milliseconds(1), fingerprintCheckTimerCallback, nullptr);
+    }).detach();
+}
+
+bool CFingerprint::isAuthenticated() {
+    return m_bAuthenticated;
+}
+
+std::optional<std::string> CFingerprint::getLastFailText() {
+    return m_sDBUSState.failText.empty() ? std::nullopt : std::optional(m_sDBUSState.failText);
+}
+
+void CFingerprint::terminate() {
+    m_sDBUSState.connection->leaveEventLoop();
+}
```

Comment:
need a check here?

Reply:
`m_sDBUSState.connection` is a nullptr when fingerprint is not enabled right?

Reply:
Done